### PR TITLE
Update cilium status command module health output

### PR DESCRIPTION
**Pull request by derailed**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

To date, we render modules health information as tabular data. As modules health instrumentation grows this rendering makes it hard for users to grok the output. This PR proposes changing the rendering of modular health using a tree like structure to give users a better line of sight to gain an understanding of the agent modules health state.

Fixes: #issue-number

```release-note
Changed cilium status CLI output to render the modules health section as a tree structure vs tabular data.
```

Signed-off-by: Fernand Galiana <fernand.galiana@gmail.com>

### CODE REVIEW

File: pkg/health/client/modules.go [hunk 1]
```diff
@@ -6,12 +6,20 @@ package client
 import (
 	"fmt"
 	"io"
+	"regexp"
 	"strings"
 
+	"github.com/opentracing/opentracing-go/log"
```
Comment (tklauser) [inline pkg/health/client/modules.go]: Any reason to use the opentracing log package for logging instead of `github.com/sirupsen/logrus` like most other cilium packages? AFAICT it's only used for the log line in `buildTree` below.

Comment (derailed) [reply pkg/health/client/modules.go]: Good catch! Thank you Tobias.

**Comment by ti-mo** [comment]
@derailed Could you update the release note in the PR description?

File: pkg/health/client/modules.go [hunk 2]
```diff
@@ -56,3 +64,49 @@ func GetAndFormatModulesHealth(w io.Writer, clt ModulesHealth, verbose bool) {
...
+		if len(ff) != 7 {
+			continue
+		}
+
+		meta := fmt.Sprintf("[%s] %s", strings.ToUpper(ff[2]), ff[5])
+		if ff[6] != noErr {
+			meta += fmt.Sprintf(" -- %s", ff[6])
+		}
+		meta += fmt.Sprintf(" (%s, x%s)", ff[3], ff[4])
+		addNode(t, mid+"."+ff[1], meta)
```
Comment (pippolo84) [inline pkg/health/client/modules.go]: The column indexes feel a bit like magic numbers, what about adding a comment with an example of the expected format and maybe some constants in place of the numbers to improve readability?

File: pkg/health/client/modules_test.go
Comment (pippolo84) [inline pkg/health/client/modules_test.go:1]: It seems that `buildTree` is not covered by the current test, might it be worth adding another one?

File: pkg/health/client/tree.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	meta interface{}
+	tree interface {
+		addNode(v val) tree
+		addNodeWithMeta(v val, m meta) tree
+		addBranch(v val) tree
+		addBranchWithMeta(v val, meta meta) tree
+		find(val val) tree
+		String() string
+		asBytes() []byte
+	}
```
Comment (pippolo84) [inline pkg/health/client/tree.go]: I don't see much value in defining these interfaces.
This tree implementation seems something specialized for the health reporter, so I'd say we can just use `string` for `val` and `meta` and simply remove the `tree` interface. This should also simplify the code a bit (e.g.: no need to use `reflect.DeepEqual` in `find` and no need for the unused `addNode`).
WDYT?

File: pkg/hive/cell/structured.go [hunk 4]
```diff
@@ -415,30 +415,39 @@ func (s *StatusNode) allOk() bool {
...
 		return ""
 	}
-	buf := bytes.NewBuffer(nil)
-	w := tabwriter.NewWriter(buf, 0, 0, 1, ' ', 0)
-	s.writeTo(w, ident)
-	w.Flush()
-	return buf.String()
+	w := bytes.NewBuffer(nil)
+	s.writeTo(w, s.Name)
+	return w.String()
```
Comment (pippolo84) [inline pkg/hive/cell/structured.go]: Nit: `strings.Builder` should be slightly more efficient and it satisfies `io.Writer` too:

```suggestion
	var sb strings.Builder
	s.writeTo(&sb, s.Name)
	return sb.String()
```

**Review by pippolo84 — COMMENTED** [review]
Great work! I'd left some questions inline.

File: pkg/health/client/tree.go (hunk 3 above)
Comment (derailed) [reply pkg/health/client/tree.go]: Good point. Thank you Fabio! Was thinking this could be useful somewhere else down the line but who knows?? I'll dumb it down.

File: pkg/health/client/modules.go (hunk 2 above)
Comment (derailed) [reply pkg/health/client/modules.go]: Thinking this will be short lived. Will update in the meantime...

File: pkg/health/client/tree.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+	leafMaxWidth            = 40
+	link         decoration = "│"
+	mid          decoration = "├──"
+	end          decoration = "└──"
+)
+
+type (
+	decoration string
+
+	// val  interface{}
```
Comment (tommyp1ckles) [inline pkg/health/client/tree.go]: Remove comments code

File: pkg/health/client/modules.go [hunk 6]
```diff
@@ -56,3 +64,55 @@ func GetAndFormatModulesHealth(w io.Writer, clt ModulesHealth, verbose bool) {
...
+		}
+		if parent != nil {
+			parent = parent.addBranch(name)
+		} else {
+			parent = p.addBranch(name)
+		}
+	}
+}
+
+func buildTree(t *node, mid, s string) {
```
Comment (tommyp1ckles) [inline pkg/health/client/modules.go]: FYI: With [#28773](https://github.com/cilium/cilium/pull/28773) I'm changing the way the module health report is sent - instead of sending a formatted string a serialized version of the tree will be sent.  I think this is the way we want to do this going forward instead of trying to parse around raw formatted text output.

I'm a bit concerned about this approach since it seems brittle (i.e. it depends on an implicit formatting of text generated by cell), so changes there will need to reflected by changes here trying to parse out the text.

Alongside the changes in https://github.com/cilium/cilium/pull/28773, could we put all this logic in one place?

The CLI text rendering code being in pkg/hive/cell is admittedly a bit awkward - perhaps we could do something like:

```go
type node struct {
    cell.StatusNode
    ...
}

func buildTree(t *node) { // note: module-id etc are now embedded in *node struct.  

}
```
So we can reuse some of that code and have all the code dealing rendering text live in the CLI code rather than in the hive/cell internals.

Wdyt?

Comment (derailed) [reply pkg/health/client/modules.go]: @tommyp1ckles Thanks! 

I was trying to operate with what was there since that pr was not on the radar as of yet. 
I'll take a peek... 

I am actually not sure we want to send out a json version of the tree??
End of day, this data will likely be managed in statedb and queried from there. 
Hence my feel is we might be best served to keep sending out an array of module health reports.

File: pkg/health/client/modules.go [hunk 7]
```diff
@@ -60,3 +68,52 @@ func GetAndFormatModulesHealth(w io.Writer, clt ModulesHealth, verbose bool) {
...
+	if err := json.Unmarshal([]byte(raw), &sn); err != nil {
+		return err
+	}
+	build(n, &sn)
+	return nil
+}
+
+func ensurePath(n *node, pp []string) *node {
+	current := n
+	for i := 0; i < len(pp); i++ {
```
Comment (tommyp1ckles) [inline pkg/health/client/modules.go]: nit: could be `i, _ := range pp { ... }`

File: pkg/health/client/modules.go [hunk 8]
```diff
@@ -60,3 +68,52 @@ func GetAndFormatModulesHealth(w io.Writer, clt ModulesHealth, verbose bool) {
...
+		if v := current.find(pp[i]); v != nil {
+			current = v
+			continue
+		}
+		name := strings.Replace(pp[i], "(/)", "", 1)
```
Comment (tommyp1ckles) [inline pkg/health/client/modules.go]: there already appears to be a constant for this (`noPod`)

File: pkg/health/client/tree.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+func (n *node) asBytes() []byte {
+	buf := new(bytes.Buffer)
+	level := 0
+	var levelsEnded []int
+	if n.parent == nil {
+		if n.meta != "" {
+			buf.WriteString(fmt.Sprintf("%v %v", n.val, n.meta))
+		} else {
+			buf.WriteString(fmt.Sprintf("%v", n.val))
```
Comment (tommyp1ckles) [inline pkg/health/client/tree.go]: nit: could be cleaner with `fmt.Fprintf`

File: pkg/health/client/modules.go [hunk 10]
```diff
@@ -32,17 +42,16 @@ func GetAndFormatModulesHealth(w io.Writer, clt ModulesHealth, verbose bool) {
...
-			n := &cell.StatusNode{}
-			if err := json.Unmarshal([]byte(m.Message), n); err != nil {
-				panic(err)
-			}
 			if m.Level == string(cell.StatusUnknown) {
 				continue
 			}
-			fmt.Fprintf(w, "%s", n.StringIndent(2))
+			if err := buildTree(r, m.Message); err != nil {
+				fmt.Fprintf(w, "Modules Health rendering failed: %s\n", err)
```
Comment (tommyp1ckles) [inline pkg/health/client/modules.go:51]: Should we do `os.Exit(1)` here?

File: pkg/health/client/modules_test.go [hunk 11]
```diff
@@ -83,13 +81,21 @@ func (m *testMHappy) GetHealth(params *daemon.GetHealthParams, opts ...daemon.Cl
 					LastUpdated: "2s",
 				},
 				{
-					ModuleID:    "m2",
-					Level:       string(cell.StatusDegraded),
-					Message:     fmt.Sprintf("{\"name\": \"m2\", \"level\":\"Degraded\",\"message\":\"doh\", \"timestamp\": %q}", t2),
+					ModuleID: "a.b.c",
+					Level:    string(cell.StatusDegraded),
+					Message: fmt.Sprintf(`
+					{"level": "OK", "name": "a.b.c", "count": 1, "sub_statuses": [
```
Comment (tommyp1ckles) [inline pkg/health/client/modules_test.go]: Could we just use `*UpdateStatus` and then marshal to make it more clear where this json comes from for someone maintaining this?

File: pkg/health/client/modules.go [hunk 12]
```diff
@@ -60,3 +69,52 @@ func GetAndFormatModulesHealth(w io.Writer, clt ModulesHealth, verbose bool) {
...
+			continue
+		}
+		name := strings.Replace(pp[i], "(/)", "", 1)
+		current = current.addBranch(name)
+	}
+
+	return current
+}
+
+func build(n *node, sn *cell.StatusNode) {
```
Comment (tommyp1ckles) [inline pkg/health/client/modules.go:95]: With this, I think the only use of `cell.(*UpdateNode).String()` family of functions is for a debug log in hive/health.go.

We should probably just get rid of that code to clean things up (doesn't have to be part of this PR - but lets make it a point to circle back and clean that up).

**Review by tommyp1ckles — COMMENTED** [review]
Output looks great! - thanks working around the last set of changes to the reporter.

Just a few comments

File: pkg/health/client/modules.go (hunk 10 above)
Comment (derailed) [reply pkg/health/client/modules.go:51]: We may not be end of line here? imho this is ok.

File: pkg/health/client/modules_test.go (hunk 11 above)
Comment (derailed) [reply pkg/health/client/modules_test.go]: Good point!

File: pkg/health/client/modules.go (hunk 12 above)
Comment (derailed) [reply pkg/health/client/modules.go:95]: Right I was wondering about that too. Don't think this is within the scope of this PR tho
Additionally tho the output has been enhance our messaging is lame. ie OK OK or worse no messages!. This need TLC and I think this should be our next priority!

File: pkg/health/client/modules.go (hunk 10 above)
Comment (tommyp1ckles) [reply pkg/health/client/modules.go:51]: Makes sense, this failing would certainly be unexpected but we can try to keep plowing forward.

File: pkg/health/client/modules.go (hunk 12 above)
Comment (tommyp1ckles) [reply pkg/health/client/modules.go:95]: Yeah right now the modular health updates are making debug logs way too noisy.  With that one we should just update the subreporter node that changed instead of formatting out the whole tree.

**Review by tommyp1ckles — APPROVED** [review]
lgmt

File: pkg/health/client/modules_test.go [hunk 13]
```diff
@@ -34,9 +35,7 @@ func TestGetAndFormatModulesHealth(t *testing.T) {
 		},
 		"happy-verbose": {
 			h: newTestMHappy(),
-			e: "Modules Health:\n" +
-				"  m1 OK status nominal 2s ago (x0)\n" +
-				"  m2 Degraded doh 20s ago (x0)",
+			e: "Modules Health:\nagent\n├── m1                                                      [OK] status nominal (2s, x0)\n└── a\n    └── b\n        └── c\n            ├── fred                                        [OK] yo (20s, x1)\n            │   └── blee                                    [OK] doh (20s, x1)\n            └── dork                                        [DEGRADED] bozo -- BOOM! (20s, x1)",
```
Comment (pippolo84) [inline pkg/health/client/modules_test.go]: Nit: using backticks here can help improve the readability of the expected output:

```go
			e: `Modules Health:
agent
├── m1                                                      [OK] status nominal (2s, x0)
└── a
    └── b
        └── c
            ├── fred                                        [OK] yo (20s, x1)
            │   └── blee                                    [OK] doh (20s, x1)
            └── dork                                        [DEGRADED] bozo -- BOOM! (20s, x1)`,
```

File: pkg/health/client/tree.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+		if l == level {
+			return true
+		}
+	}
+
+	return false
+}
+
+func dumpVal(level int, node *node) string {
+	lines := strings.Split(fmt.Sprintf("%v", node.val), "\n")
```
Comment (pippolo84) [inline pkg/health/client/tree.go]: Now that `val` is a string this can be simplified:

```go
lines := strings.Split(node.val, "\n")
```

There are also some other places where `%s` can be used instead of `%v`

**Review by pippolo84 — APPROVED** [review]
LGTM, thanks! :rocket: 
A couple of non-blocking nits left inline.

File: pkg/health/client/tree.go (hunk 14 above)
Comment (derailed) [reply pkg/health/client/tree.go]: Great catch! Tx Fabio I'll update.

**Review by jrajahalme — APPROVED** [review]

**Review by tklauser — APPROVED** [review]
